### PR TITLE
Singularity Shaders and a lot of Shader Stuff 

### PR DESCRIPTION
**Pull request by GraniteSidewalk**
Requires space-wizards/RobustToolbox#1387 
Ready to merge(?)
This does:
- Adds a variety of shaders: singularity shader (gravitational lensing effect), texture shader, crit state shader (dark circle around your character), glowing red border (also for crit state)
- Changes a bunch of item descriptions since I dunno I saw them and felt like it
- Singulo changes

Fixes https://github.com/space-wizards/space-station-14/issues/2212

### CODE REVIEW

**Comment by GlassEclipse** [comment]
Examples of what it currently looks like. Do give feedback on what you think could be improved.
https://imgur.com/a/ESUvXOI
Resolution is slightly low (due to imgur I think).

**Comment by AJCM-git** [comment]
How does it look whitout the sprite?

**Comment by GlassEclipse** [comment]
Not very good, mostly since the shader I've created is edited to try and look nice with the singulo in the middle (thus removing most of the reflection in the center). I've still recorded it anyway, if we wanted to go that route it would then look like a really wacky clown mirror in the middle.
Here is the no sprite version and a higher res of the normal version: https://imgur.com/a/zqqxjPy

**Comment by brndd** [comment]
I know it's not how real black holes work, but I think a simple "pinch" effect would work too and maybe even look better the current effect. The current effect acts like a magnifying glass, so it feels like things are being pushed away from the black hole rather than being sucked into it. A "pinch" effect gives the opposite feel, at least to me, even though it's not realistic.

pinch

Could perhaps apply a "whirl" effect too to go with the singulo's rotation:

whirl

**Comment by GlassEclipse** [comment]
I can try that.

**Comment by brndd** [comment]
Yeah, I realize I'm asking you to essentially do the exact opposite effect of the effect you worked a long time to make, so it sounds a bit silly... But I feel like in cartoon physics a "pinch" is more intuitive, as it gives the effect of things stretching *towards* the black hole as they're sucked into it.

**Comment by GlassEclipse** [comment]
Nah, I think I can just inverse the value on the distortion factor.

**Comment by GlassEclipse** [comment]
This is what it looks like when I make an addition operation subtraction instead. Effect seems extreme since the values are tuned for the outward effect, but those can be changed very easily. Freezes at the end since the singularity tries to move root node.
https://imgur.com/a/x0DEGqf

**Comment by GlassEclipse** [comment]
Also, swirl effect will require more effort so I'll only work on it if you think it'll look better.

**Comment by brndd** [comment]
I think that looks really good. I think it's what a cartoon physics black hole should look like with things stretching into it like that. The effect strength probably should be tweaked as you said.

Is the gravitational lensing effect where things behind the black hole turn into a ring around it mutually exclusive with this? That's the one thing I'm missing from the more realistic effect, but logically thinking I'm not sure how that would mix with this.

**Comment by metalgearsloth** [comment]
I think I prefer glass's? I'd need to see the pinch in motion to get a better comparison. Also @GlassEclipse chuck it on progress-reports.

**Comment by GlassEclipse** [comment]
To be honest I have no idea if the circular lensing effect can be somehow emulated. I don't think it can, since the inward effect condenses textures into an inward circle instead of an outward circle. But maybe it can reconfigured so that the inward circle is on the rim of the singularity texture? Sounds complicated either way.
This is what it looks like without the singularity texture on top, so that's what I'm working with here. Probably easier to just do the swirl effect.
Screenshot 2020-11-08 22 23 36

**Comment by brndd** [comment]
Well, either way I really like the pinching effect. It was stated on Discord that it also makes the "danger zone" around the singularity a bit more evident.

**Comment by GlassEclipse** [comment]
Well personally I think the original effect looks better, plus I already did the fine tuning for that, so I'm going to stick with it. It's really easy to switch it back to inward suck, so we can if we so desire. 
I'll be fixing some bugs I found then should be ready for merge if PJB thinks I didn't totally slow down the rendering code by 5x with my constant framebuffer copying and render queue flushing.

**Comment by GlassEclipse** [comment]
Should be ready to merge pending code review. Fixed all the bugs I can see, I am sure there will be more but I will fix them as they come up.

**Comment by GlassEclipse** [comment]
10 years since last merge.i walk through the empty streets trying to think of something else but my path always leads to merging prs. i stare at the screen for hours and try to summon the lord pjb. i watch other prs get merged and fix conflicts but it is no good. tonight i cry myself to sleep

**Comment by GraniteSidewalk** [comment]
As a status update, the engine PR is almost finished. I've had some irl stuff that I thought would be over now but isn't so it may still be a while yet. I will still need to change the content PR too.

**Review by PaulRitter — CHANGES_REQUESTED** [review]
Files i'd like to see `#nullable enable` at the start of:
- Content.Client/GameObjects/Components/Mobs/ClientOverlayEffectsComponent.cs
- Content.Client/GameObjects/Components/Suspicion/SuspicionRoleComponent.cs
- Content.Client/GameObjects/Components/Weapons/FlashableComponent.cs
- Content.Client/Graphics/Overlays/ColoredScreenBorderOverlay.cs
- Content.Client/Graphics/Overlays/SingularityOverlay.cs
- Content.Client/Graphics/Overlays/TextureOverlay.cs
- Content.Server/GameObjects/Components/Effects/BaseShaderAuraComponent.cs
- Content.Server/GameObjects/Components/Effects/SingularityShaderAuraComponent.cs
- Content.Server/GameObjects/Components/Mobs/ServerOverlayEffectsComponent.cs
- Content.Server/GameObjects/Components/Weapon/Melee/FlashComponent.cs
- Content.Server/GameObjects/EntitySystems/BaseShaderAuraTickSystem.cs
- Content.Server/Mobs/Commands.cs
- Content.Shared/GameObjects/Components/Mobs/SharedOverlayParameters.cs

File: Content.Server/Commands/Mobs/RemoveOverlayCommand.cs [hunk 1]
```diff
@@ -9,9 +9,9 @@ namespace Content.Server.Commands.Mobs
     [AdminCommand(AdminFlags.Debug)]
     public class RemoveOverlayCommand : IClientCommand
     {
-        public string Command => "rmoverlay";
-        public string Description => "Removes an overlay by its ID";
-        public string Help => "rmoverlay <id>";
+        public string Command => "rmoverlays";
+        public string Description => "Removes all overlays of a given type.";
+        public string Help => "rmoverlays <id>";
```
Comment (Visne) [inline Content.Server/Commands/Mobs/RemoveOverlayCommand.cs:14]: Seems like `<id>` should be changed here.

File: Content.Server/Mobs/Commands.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+using Robust.Server.Interfaces.Player;
+using Robust.Shared.IoC;
+using Robust.Shared.Network;
+using Robust.Shared.Prototypes;
+
+namespace Content.Server.Mobs
+{
+
+    [AdminCommand(AdminFlags.Admin)]
+    public class MindInfoCommand : IClientCommand
```
Comment (Visne) [inline Content.Server/Mobs/Commands.cs:20]: These should probably be in seperate files.

File: Resources/Prototypes/Entities/Objects/Fun/toys.yml [hunk 3]
```diff
@@ -129,7 +129,7 @@
   name: help me carving
   parent: BaseItem
   id: CarvingHelpMe
-  description: Help me...
+  description: Help me
```
Comment (Visne) [inline Resources/Prototypes/Entities/Objects/Fun/toys.yml:132]: In my opinion there should be at least one period.

Comment (GraniteSidewalk) [reply Resources/Prototypes/Entities/Objects/Fun/toys.yml:132]: There's actually an exclamation (according to the actual dark souls desc)

**Pull request merged**